{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[{"file_id":"1fEuAoXMORAHdfgN2FXiV7pzIDWTtvUs1","timestamp":1678333441273}]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["#Background                                                   \n","In this code, we will write a program to detect ORFs in a given input nucleotide sequence. We will start by writing an algorithm based on our understanding of ORFs."],"metadata":{"id":"cxowaEb1huvd"}},{"cell_type":"markdown","source":["#Algorithm 1\n","\n","\n","1.   Ask the person/user to input their genome sequence.\n","2.   Take the Amino acid codon table.\n","3.   Convert all the letter \"U\" in codon table to \"T.\"\n","4.   Declare the sequence \"ATG\" as a start codon. And \"TAG, TAA and TGA\" as stop codons.\n","5.   Start reading the genome sequence from left to write and call it as \"sense strand\" and in display store value as \"+\".\n","6.   Make and output table with columns ; [ORF number/gene number,         sequence, gene length, and strand direction]\n","7.   Record the sense strand reading into the column- strand direction.\n","8.   Read/match the sense strand one one letter each to find the start codon. \n","9.   Until you detect the start codon keep reading the strand. \n","10.  Read the start codon and Keep the tract of the codon number or nucleotide position.     \n","11.  The position you find the start codon, divide the further sequence in triplet. This triplet are named as codon.\n","12.  Compare the codon sequence from the codon table and assigned the associated single letter code of amino acid.\n","13.  In sequence codon there will be two options to store values.               \n","      a. Nucleotide sequence.                                     \n","      b. Protein sequence            \n","14.  In nucleotide sequence do not convert the codon bases to amino acid single letter code instead just store whatever letters/bases are there after start codon into the column of sequence in nucleotide sequence.\n","15.  For protein sequence, follow the step 12 and store the letter value in protein sequence column.\n","16.  Keep storing the values and reading the sequence in trio's until you tackle the stop codon.\n","17.  Read the stop codon and keep the track of the stop codon or nucleotide position.      \n","18.  Stop dividing the sequence in three.\n","19.  Call this sequence from start codon to stop codon as a ORF or a gene.\n","20.  Store the value of the ORF number in the column of ORF number. \n","21.  Number the ORF as ORF x where is a belongs to natural number.\n","22.  Again continue matching the sequence as one letter to another after the previous stop codon ended.This is in assumption that we have to ignore nesting and overlapping ORFs. \n","23.  Find the start codon and again repeat the steps 5-20.\n","24.  Take the previous ORF number as x and add 1 to it.\n","25.  Number the orf as ORF (x+1) and store it as a new entry in the followed row.\n","26.  After finishing the entire genome sequence from left to write, Reverse the sequence reading from right to left and call it as \"anti-strand\" and store it's value as \"-\" in the strand direction column.\n","27.  Repeat all the above 8-24 steps for reverse strand.\n","28.  To include Overlapping ORFs, again read the sequence from left to right letter by letter and ignore all the start codons which are already read as ORFs and move forward to find the next start codon.\n","29.  take the next unread start codon and keep the nucleotide number saved.\n","30.  Repeat all the above steps till we encounter the stop codon.\n","31.  Ignore the already read stop codon and take the next unread stop codon and mark it as another ORF. \n","32.  For nested ORF \n"],"metadata":{"id":"9oCUCyagn-nY"}},{"cell_type":"markdown","source":["#Algorithm 2                          \n","\n","\n","1.   Give the input nucleotide/genome sequence.\n","2.   Insert the amino acid codon table.\n","3.   Save start codon as\"ATG\" and stop codons as \"TAA, TAG and TGA.\"\n","4.   Ask the program to start searching for start codon starting from the left of the sequence moving towards right. Call this is a sense strant and show as \"+\" \n","5.   The moment it detects the ATG/start sequence start dividing the further bases into triplets and look for any of the stop codon.\n","6.   Stop dividing the bases into triplet.\n","7.   Call the sequence starting from start codon to stop codon as gene x or orf x. [here x= N; N= natural number]\n","8.   While dividing the codons, there will be many start codon in between the start and stop, do not take a start codon in the middle as a new gene entry instead only consider genes those starts with start codon and ends with stop. Ignore the nested genes. \n","9.   Again start searching for start codon in right direction and follow the steps 4-7 and name the gene number as gene (x+1).\n","10.  Repeat the above process till the sequence ends.\n","11.  Ask the program to for cut off.\n","12.  There will be 5 options for cutoff; 30, 100, 200, 300, 500.\n","13.  If any of the cut off is choosen, verify the gene length and only give results of those sequence who are equal to or greater than the chose cutoff.\n","14.     \n"],"metadata":{"id":"a6E5C70Kh_Ox"}},{"cell_type":"markdown","source":["#Final Algorithm \n","\n","\"CATGCATGCATGAGTCGATCAGTCAGTGAGCGATGCATCAGTC\"\n","\n","1. Open the seq to be able to read.\n","2. number the nucleotides\n","3. define/open the universal codon table\n","4. break the seq into pair of 3nts starting from 1\n","5. make it frame 1\n","6. start matching with the start codon data\n","7. keep track of the nts number\n","8. if match found \n","    record the position nts number\n","    if match not found move to next frame\n","9. start matching with the stop codon data\n","10. if match found for stop codon                  \n","    record the nts number and orf number               \n","    take the seq between start & stop number                        \n","    repeat the step 6 to 9                   \n","11. if match not found for stop codon           \n","    move to next frame                  \n","    break the seq into 3 from nts number 2              \n","    name this new frame 2                  \n","12. Repeat from 6 to 10 for frame 2\n","13. Repeat from 6 to 10 for frame 3"],"metadata":{"id":"jZRi3RdO-g7b"}},{"cell_type":"markdown","source":["#pseudo code\n"],"metadata":{"id":"Cg2TBAISblYi"}},{"cell_type":"markdown","source":["#Defining variables"],"metadata":{"id":"viZFJxESb7Gs"}},{"cell_type":"code","source":["#using Biopython library\n","\n","!pip install Bio\n","from Bio import SeqIO   #using sequence function of the Bio library to access the file seq data.\n","\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"7wylO_OS03sc","executionInfo":{"status":"ok","timestamp":1678643315755,"user_tz":-330,"elapsed":9930,"user":{"displayName":"SAKSHI TEMBHURNE","userId":"05875301864129969279"}},"outputId":"3cc93e69-f90e-4bc3-8e3f-3890282482df"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n","Collecting Bio\n","  Downloading bio-1.5.6-py3-none-any.whl (275 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m275.6/275.6 KB\u001b[0m \u001b[31m5.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: pandas in /usr/local/lib/python3.9/dist-packages (from Bio) (1.3.5)\n","Requirement already satisfied: tqdm in /usr/local/lib/python3.9/dist-packages (from Bio) (4.65.0)\n","Requirement already satisfied: pooch in /usr/local/lib/python3.9/dist-packages (from Bio) (1.7.0)\n","Collecting mygene\n","  Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n","Collecting gprofiler-official\n","  Downloading gprofiler_official-1.0.0-py3-none-any.whl (9.3 kB)\n","Requirement already satisfied: requests in /usr/local/lib/python3.9/dist-packages (from Bio) (2.25.1)\n","Collecting biopython>=1.80\n","  Downloading biopython-1.81-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m36.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.9/dist-packages (from biopython>=1.80->Bio) (1.22.4)\n","Collecting biothings-client>=0.2.6\n","  Downloading biothings_client-0.2.6-py2.py3-none-any.whl (37 kB)\n","Requirement already satisfied: python-dateutil>=2.7.3 in /usr/local/lib/python3.9/dist-packages (from pandas->Bio) (2.8.2)\n","Requirement already satisfied: pytz>=2017.3 in /usr/local/lib/python3.9/dist-packages (from pandas->Bio) (2022.7.1)\n","Requirement already satisfied: platformdirs>=2.5.0 in /usr/local/lib/python3.9/dist-packages (from pooch->Bio) (3.1.0)\n","Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.9/dist-packages (from pooch->Bio) (23.0)\n","Requirement already satisfied: chardet<5,>=3.0.2 in /usr/local/lib/python3.9/dist-packages (from requests->Bio) (4.0.0)\n","Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.9/dist-packages (from requests->Bio) (2.10)\n","Requirement already satisfied: urllib3<1.27,>=1.21.1 in /usr/local/lib/python3.9/dist-packages (from requests->Bio) (1.26.14)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.9/dist-packages (from requests->Bio) (2022.12.7)\n","Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.9/dist-packages (from python-dateutil>=2.7.3->pandas->Bio) (1.15.0)\n","Installing collected packages: biopython, gprofiler-official, biothings-client, mygene, Bio\n","Successfully installed Bio-1.5.6 biopython-1.81 biothings-client-0.2.6 gprofiler-official-1.0.0 mygene-3.2.2\n"]}]},{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/drive')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"o7fyRGr141aS","executionInfo":{"status":"ok","timestamp":1678643354148,"user_tz":-330,"elapsed":35366,"user":{"displayName":"SAKSHI TEMBHURNE","userId":"05875301864129969279"}},"outputId":"88da2bdb-283b-4947-d54c-9a9beb33255c"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/drive\n"]}]},{"cell_type":"code","source":["\n","record = SeqIO.read(\"MS2_phage_seq.fasta\", \"fasta\")\n","print(record.seq)       #record.seq will only print the seq part of the file."],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":329},"id":"xlKt-E1L2VpI","executionInfo":{"status":"error","timestamp":1678643365925,"user_tz":-330,"elapsed":669,"user":{"displayName":"SAKSHI TEMBHURNE","userId":"05875301864129969279"}},"outputId":"ffe892b9-8ac1-47d4-d068-fcf2f1043f65"},"execution_count":null,"outputs":[{"output_type":"error","ename":"FileNotFoundError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-5-960b8b29e58f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mrecord\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"MS2_phage_seq.fasta\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrecord\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.9/dist-packages/Bio/SeqIO/__init__.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(handle, format, alphabet)\u001b[0m\n\u001b[1;32m    650\u001b[0m     \u001b[0mto\u001b[0m \u001b[0mread\u001b[0m \u001b[0mmultiple\u001b[0m \u001b[0mrecords\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mhandle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    651\u001b[0m     \"\"\"\n\u001b[0;32m--> 652\u001b[0;31m     \u001b[0miterator\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mformat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0malphabet\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    653\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    654\u001b[0m         \u001b[0mrecord\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0miterator\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.9/dist-packages/Bio/SeqIO/__init__.py\u001b[0m in \u001b[0;36mparse\u001b[0;34m(handle, format, alphabet)\u001b[0m\n\u001b[1;32m    603\u001b[0m     \u001b[0miterator_generator\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_FormatToIterator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    604\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0miterator_generator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 605\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0miterator_generator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    606\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mformat\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mAlignIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_FormatToIterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    607\u001b[0m         \u001b[0;31m# Use Bio.AlignIO to read in the alignments\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.9/dist-packages/Bio/SeqIO/FastaIO.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, source, alphabet, title2ids)\u001b[0m\n\u001b[1;32m    221\u001b[0m             )\n\u001b[1;32m    222\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtitle2ids\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtitle2ids\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 223\u001b[0;31m         \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msource\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"t\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfmt\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"Fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    224\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    225\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhandle\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.9/dist-packages/Bio/SeqIO/Interfaces.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, source, alphabet, mode, fmt)\u001b[0m\n\u001b[1;32m     43\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"The alphabet argument is no longer supported\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     44\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 45\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstream\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msource\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"r\"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     46\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshould_close_stream\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     47\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# not a path, assume we received a stream\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'MS2_phage_seq.fasta'"]}]},{"cell_type":"code","source":["# defining variables\n","input_seq = record.seq              #using the seq of the file as a input seq for further processing.\n","start_codon = \"ATG\"\n","stop_codons = [\"TAG\",\"TGA\",\"TAA\"]\n","ORF_num = 0\n","frame_num = 1\n","ORF_begin_loc = 0\n","ORF_end_loc = 0\n","ORF_len = 75                                                                     #subject to change \n","\n","#making empty dictionary for storing ORF sequences. \n","ORF_seqs = {}\n","\n","#dividing the sequence into 3 by creating subsets of a string.\n","nt_pos=0          #initial value is 0.\n","while nt_pos <(len(input_seq)-2):\n","##for nt_pos in range(0,(len(input_seq)-2),3):                                  #range(start,stop,step)\n","                                              \n","  read_codon=input_seq[nt_pos: nt_pos+3]                                        #storing the values into another temporary variablea\n","  if read_codon==start_codon:                                                   #stating the condition that the read codon is equal to the start codon or not, if it is proceed further.\n","    for new_pos in range(nt_pos+3,(len(input_seq)-2),3):                        #storing the new values\n","      read_codon2=input_seq[new_pos:new_pos+3]\n","      if read_codon2 in stop_codons:\n","        if (new_pos-nt_pos) >= ORF_len:       #adding length cut off for orf sequence. Sequences of length greater than or equal to the ORF_len will only be selected\n","          ORF_begin_loc=nt_pos                #storing orf begin and end locations in the new introduced variables\n","          ORF_end_loc=new_pos\n","          ORF_num= ORF_num+1                  #increasing the orf number by one in each cycle. \n","          #ORF_seq=input_seq[ORF_begin_loc:(ORF_end_loc+3)] #here the orf_seq which was dict has now chnaged to a string as we are storing all the for loop values in it.\n","          ORF_seqs[\"ORF\"+str(ORF_num)]=input_seq[ORF_begin_loc:(ORF_end_loc+3)] #storing values directly to the dict ##[\"ORF\"+str(ORF_num)]is the key and input_seq[ORF_begin_loc:(ORF_end_loc+3)] is the value.\n","          nt_pos=new_pos+3                    # to prevent loop to again start with the same location we are saying that for every cycle add three to the start position location.  \n","        break                                  # if we find a stop codon then we need to get out of the loop thus use break to break the loop and not continue the process.    \n","\n","  nt_pos=nt_pos+3                              # even if we don't find the stop codon we still want it to restart again.\n","\n","# printing the orf begin and end location, orf number and orf sequences.                                           \n","print(ORF_begin_loc)\n","print(ORF_end_loc)\n","print(ORF_num)\n","#print(ORF_seq)\n","print(ORF_seqs)\n","\n","    \n","\n"],"metadata":{"id":"F-0cuwPJbtrL","executionInfo":{"status":"ok","timestamp":1678430860803,"user_tz":-330,"elapsed":522,"user":{"displayName":"SAKSHI TEMBHURNE","userId":"05875301864129969279"}},"colab":{"base_uri":"https://localhost:8080/"},"outputId":"6cca4090-0244-4af5-d144-16afda3c9a35"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["2037\n","2235\n","3\n","{'ORF1': Seq('ATGAGTGATATCCAGGGTGCATATGAGATGCTTACGAAGGTTCACCTTCAAGAG...TAG'), 'ORF2': Seq('ATGGAAACCCGATTCCCTCAGCAATCGCAGCAAACTCCGGCATCTACTAATAGA...TAA'), 'ORF3': Seq('ATGACAAATCCTTGTCATGGGATCCGGATGTTTTACAAACCAGCATCCGTAGCC...TGA')}\n"]}]}]}