{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyMf0cRwwU8Ib6foOV+e30Hf"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":null,"metadata":{"id":"5RD1veHbkUoA"},"outputs":[],"source":[]},{"cell_type":"markdown","source":["#Algorithm\n","1. Input the Template sequence from 5' to 3' direction.         \n","2. Input the start and end location of the region to be amplified. \n","3. The primer length will be 20 amino acids long.         \n","4. To form the primer give the complementary of the bases that is A compliments to T and G compliments to C and vice versa.\n","5. To find a good primer sequence we need to check for GC% in the primer sequence. \n","6. The primer sequence of length 20 has to have a total number of G and C in all the four bases; A,T,G and C in between 0.45 to 0.55. If this condition is satisfied then we can allow the primer to be formed.\n","7. Search for primer before the start position and after the end poaition of the required amplified region.  \n","The search for primer from the start and end position of the region to be amplified in opposite direction. \n","8. Ask user to input the range where we can look for primer sequence. \n","9. Search in opposite direction of start and end direction in a stretch of 20 aa. \n","10. Verify the GC%, if it lies in the range then allow it. Record the positions.   \n","11. If not then move one letter ahead in a stretch of 20 aa and verify the GC%.\n","12. Even after finding a primer which is fulfiling the condition continue looking for other primers till the condition/allowed range given by the user. Record its positions.\n","13. Rate all the primers found according to the their GC%. \n","14. Shortlist the top 10 primers.\n","15. For primer search started from start position; arrange the primers in ascending order of their position to keep the orientation as 5' to 3'. \n","16. For primer search started from end position; arrange the primers in decending order of their position to keep the orientation as 5' to 3'.\n","17. Suggest top 5 which are most close to the region to be amplified after doing the shortlisting for both the start and end direction separately.\n"],"metadata":{"id":"Wnq5TjvtBfuz"}},{"cell_type":"markdown","source":["#Final Algorithm \n","primer can be a part of template. Output table - Locatoin, Sequence, Orientation, GC percentage and number Processivity - different enzymes have differnet strenght to process a certin number of base pairs.\n","\n","1. Take input sequence in 5' to 3'direction.                \n","2. Ask for range of region to be amplified.                  \n","3. Specify maximum allowed buffer range for primer.                         \n","4. length of buffer to be taken as 20 which is constant.\n","similarly primer 2\n","\n","For forward primer\n","5. Read sequence from s(start) to s + 19.\n","6. Find ATCG , calculate the GC%.\n","7. Check if GC% is wihtin allowed range.\n","8. if  yes, then output the nucleotide sequence, location/position, GC%\n","9. Read from (S-1) ----> (S-1+19)\n","10. Repeat steps 5-7\n","11. If reached till S-200\n","12. Break\n","13. Sort the data in GC%. \n","\n","For Reverse primer\n","1. Make the input seq complementary.\n","2. Read the c.seq from e-19---> e.\n","3. Find ATCG , calculate the GC%.\n","4. Check if GC% is wihtin allowed range.\n","5. If yes, then output the nucleotide sequence, location/position as (E -19) from 5' to 3' - so reverse the sequence, GC%.\n","6. (e-19+1)----->(e+1)\n","7. if reached (e+200-20)\n"],"metadata":{"id":"GtshtGeNJ6UR"}},{"cell_type":"code","source":["input_seq=\"ATAAGACTCCTTAAGTTCTCTACAATGTACGACCCGGAAAGGTTGCACAACTTAGACCCTACAATGTACTAGTAGGCCTCTGCGCTGGCAATACAGATAAGATAATGTAGCCCCTGGCCTCAAAGGAACTCTCCTCCTTAGGTTGCATTTGTATAATGTTTGATTTTTAGAACAAAGGTTGCATGTACTAGTTCTCCTCCTTAAGTTCTCGTTCTCCTCCAATAAGAATAAGATAATGTT\"\n","pcr_start_loc = s\n","pcr_start_loc = 30\n","pcr_end_loc = e\n","pcr_end_loc = 80\n","primer_len = 20\n","buffer_range =10\n","GC_pct = 0.45\n","\n","primer_pos = s\n","primer_range=input_seq[s : s+(primer_len-1)]\n","if primer_range==GC_pct\n","  then\n","  for primer_pos1 in range [(s-1):(s-1)+(primer_len-1)]:                       #(S-1) ----> (S-1+19)\n","    primer_range2=input_seq[(s-2):(s-2)+(primer_len-1)]\n","\n","    pri_begin_loc=primer_pos\n","    pri_end_loc=primer_pos1\n","\n","\n","while nt_pos <(len(input_seq)-2):\n","##for nt_pos in range(0,(len(input_seq)-2),3):                                  #range(start,stop,step)\n","                                                  #input_seq[x,x+3]/[0:3]\n","  read_codon=input_seq[nt_pos: nt_pos+3]                                        #storing the values into another temporary variablea\n","  if read_codon==start_codon:\n","    for new_pos in range(nt_pos+3,(len(input_seq)-2),3):\n","      read_codon2=input_seq[new_pos:new_pos+3]\n","      if read_codon2 in stop_codons:\n","        if (new_pos-nt_pos) >= ORF_len:       #adding length cut off for orf sequence. \n","          ORF_begin_loc=nt_pos                #storing orf begin and end locations in the new introduced variables\n","          ORF_end_loc=new_pos\n","          ORF_num= ORF_num+1\n","          #ORF_seq=input_seq[ORF_begin_loc:(ORF_end_loc+3)] #here the orf_seq which was dict has now chnaged to a string as we are storing all the for loop values in it.\n","          ORF_seqs[\"ORF\"+str(ORF_num)]=input_seq[ORF_begin_loc:(ORF_end_loc+3)] #storing values directly to the dict ##[\"ORF\"+str(ORF_num)]is the key and input_seq[ORF_begin_loc:(ORF_end_loc+3)] is the value.\n","          nt_pos=new_pos+3\n","        break\n","\n"],"metadata":{"id":"tQL2X74vRNLI"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["\n","\n","def GC_ratio(my_seq):  #function to calculate GC #Why are we using another variable my_seq and not input_seq?\n","  count_A=my_seq.count(\"A\") #count every nucleotide - A, T, G and C #count function from my_seq and count for a string A \n","  count_T=my_seq.count(\"T\") #A and T do not need to be counted as we know GC and the length of the sequence. \n","  count_G=my_seq.count(\"G\")\n","  count_C=my_seq.count(\"C\")     \n","  GC_per = 100*((count_G + count_C)/(len(my_seq))) \n","  return(GC_per) #return tells python to give the output = value of GC percentage \n","\n","\n","  #This code is used to find the forward primer sequence that lies within a given start and stop region\n","#Input \n","input_seq = \"ATAAGACTCCTTAAGTTCTCTACAATGTACGACCCGGAAAGGTTGCACAACTTAGACCCTACAATGTACTAGTAGGCCTCTGCGCTGGCAATACAGATAAGATAATGTAGCCCCTGGCCTCAAAGGAACTCTCCTCCTTAGGTTGCATTTGTATAATGTTTGATTTTTAGAACAAAGGTTGCATGTACTAGTTCTCCTCCTTAAGTTCTCGTTCTCCTCCAATAAGAATAAGATAATGTT\"\n","primer_len = 20                              #primer len is set as 20 but it is a input which user will provide\n","pcr_start_loc = 29                      #asking user to input the start and end position for the region to be amplified.\n","stop_region =79\n","buffer_range = 10                         #it is the allowed range to search for the primers on either side of region to be amplified.\n","\n","#For the forward primer \n","for primer_start in range(pcr_start_loc,(pcr_start_loc-buffer_range),-1):  \n","  primer_seq = (input_seq[primer_start:(primer_start+primer_len)]) #Don't need to do - 1, beacuse S---->S+19 will give 20 nucleotide sequence. We also made primer_len as variable and not just number so that if we have to change the value of it, we don't need to change the code. \n","  #print(primer_range) \n","\n","  GC_ratio(primer_seq)           #calculates the percentage of GC from the primer found - which is saved as primer_range, the value is coming from the function in the previous code of def If the GC percentage falls within the range of 45-55%, we consider this primer sequence to be valid\n","  if 45 < GC_ratio(primer_seq) < 55:                                 #checking the gc ratio is in given range \n","    print(primer_seq)\n","    print(primer_start)\n","    print(GC_ratio(primer_seq))                           #This code is used to find the forward primer sequence that lies within a given start and stop region \n","\n","  "],"metadata":{"id":"Le5Ur0kQ_cxZ","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1679122194884,"user_tz":-330,"elapsed":515,"user":{"displayName":"SAKSHI TEMBHURNE","userId":"05875301864129969279"}},"outputId":"c44aa0d5-8bfd-40ef-d67e-717dbaa46359"},"execution_count":7,"outputs":[{"output_type":"stream","name":"stdout","text":["ATGTACGACCCGGAAAGGTT\n","24\n","50.0\n","AATGTACGACCCGGAAAGGT\n","23\n","50.0\n","ACAATGTACGACCCGGAAAG\n","21\n","50.0\n"]}]},{"cell_type":"code","source":["#modified forward primer program \n","def Tm(my_seq):             #defining melting temperature\n","  count_A=my_seq.count(\"A\") #count every nucleotide - A, T, G and C #count function from my_seq and count for a string A \n","  count_T=my_seq.count(\"T\") \n","  count_G=my_seq.count(\"G\")\n","  count_C=my_seq.count(\"C\")     \n","  temp = (4*(count_G + count_C))+ (2*(count_A + count_T))\n","  return temp                      #return tells python to give the output = value of melting temperature\n","\n","\n","  #This code is used to find the forward primer sequence that lies within a given start and stop region\n","#Input \n","input_seq = \"ATAAGACTCCTTAAGTTCTCTACAATGTACGACCCGGAAAGGTTGCACAACTTAGACCCTACAATGTACTAGTAGGCCTCTGCGCTGGCAATACAGATAAGATAATGTAGCCCCTGGCCTCAAAGGAACTCTCCTCCTTAGGTTGCATTTGTATAATGTTTGATTTTTAGAACAAAGGTTGCATGTACTAGTTCTCCTCCTTAAGTTCTCGTTCTCCTCCAATAAGAATAAGATAATGTT\"\n","primer_len = 20                              #primer len is set as 20 but it is a input which user will provide\n","pcr_start_loc = 29                           #asking user to input the start and end position for the region to be amplified.\n","stop_region =79\n","buffer_range = 10                         #it is the allowed range to search for the primers on either side of region to be amplified.\n","\n","#For the forward primer \n","for primer_start in range(pcr_start_loc,(pcr_start_loc-buffer_range),-1):  \n","  primer_seq = (input_seq[primer_start:(primer_start+primer_len)])  #searching for primer sequence in this given range.\n","  #print(primer_range) \n","  tm = Tm(primer_seq) \n"," # print(tm)          \n","  if 50 < tm < 65:                                 #checking the tm is in given range \n","    print(primer_seq)\n","    print(primer_start)\n","    print(Tm(primer_seq))                           #This code is used to find the forward primer sequence that lies within a given start and stop region \n","\n","\n","  "],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Kk3TKpFac-RQ","executionInfo":{"status":"ok","timestamp":1679122891638,"user_tz":-330,"elapsed":403,"user":{"displayName":"SAKSHI TEMBHURNE","userId":"05875301864129969279"}},"outputId":"e49a85c9-f171-4010-cc5f-706870f07d24"},"execution_count":15,"outputs":[{"output_type":"stream","name":"stdout","text":["CGACCCGGAAAGGTTGCACA\n","29\n","64\n","ACGACCCGGAAAGGTTGCAC\n","28\n","64\n","TACGACCCGGAAAGGTTGCA\n","27\n","62\n","GTACGACCCGGAAAGGTTGC\n","26\n","64\n","TGTACGACCCGGAAAGGTTG\n","25\n","62\n","ATGTACGACCCGGAAAGGTT\n","24\n","60\n","AATGTACGACCCGGAAAGGT\n","23\n","60\n","CAATGTACGACCCGGAAAGG\n","22\n","62\n","ACAATGTACGACCCGGAAAG\n","21\n","60\n","TACAATGTACGACCCGGAAA\n","20\n","58\n"]}]}]}